llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp

```diff
@@ -1354,6 +1358,40 @@ SDValue LoongArchTargetLowering::lowerVECTOR_SHUFFLE(SDValue Op,
   return SDValue();
 }
 
+SDValue LoongArchTargetLowering::lowerFP_TO_FP16(SDValue Op,
+                                                 SelectionDAG &DAG) const {
+  // Custom lower to ensure the libcall return is passed in an FPR on hard
+  // float ABIs.
+  SDLoc DL(Op);
+  MakeLibCallOptions CallOptions;
+  SDValue Op0 = Op.getOperand(0);
+  SDValue Chain = SDValue();
+  RTLIB::Libcall LC = RTLIB::getFPROUND(Op0.getValueType(), MVT::f16);
+  SDValue Res;
+  std::tie(Res, Chain) =
+      makeLibCall(DAG, LC, MVT::f32, Op0, CallOptions, DL, Chain);
+  if (Subtarget.is64Bit())
+    return DAG.getNode(LoongArchISD::MOVFR2GR_S_LA64, DL, MVT::i64, Res);
+  return DAG.getBitcast(MVT::i32, Res);
+}
+
+SDValue LoongArchTargetLowering::lowerFP16_TO_FP(SDValue Op,
+                                                 SelectionDAG &DAG) const {
+  // Custom lower to ensure the libcall argument is passed in an FPR on hard
+  // float ABIs.
```

Comment:
You shouldn't need to custom lower the casts to change the ABI. Are you trying to special case the ABI for this one call in this one instance? That seems bad

Reply:
We're not trying to special case the ABI, but rather ensuring compliance with the ABI rules for floating-point operations. Specifically, the argument for `f32 __gnu_h2f_ieee(f16)` needs to be passed via the FPR, as per the floating-point ABI, rather than the GPR. Custom lowering ensures that the argument is correctly passed through the FPR in cases where the default behavior doesn't align with this requirement. Thanks.

Reply:
This change does not accomplish this. The cast opcodes have nothing to do with the ABI, other than ABI code may result in inserting them

Reply:
The default behavior of `softPromoteHalf` does not align with the expectations of the architecture, the custom lowering as referenced by the approach used in RISC-V.

Ref: https://reviews.llvm.org/D151284
> This legalisation produces ISD::FP_TO_FP16 and ISD::FP16_TO_FP nodes which (as described in ISDOpcodes.h) provide a "semi-softened interface for dealing with f16 (as an i16)". i.e. the return type of the FP_TO_FP16 is an integer rather than a float (and the arg of FP16_TO_FP is an integer). The remainder of the description focuses primarily on FP_TO_FP16 for ease of explanation.

Is there a better approach to achieve this?

Reply:
In other words, you are special casing the ABI for this one libcall that happens to be used for legalization of the conversions. If you are fixing the ABI, as the title suggests, you need to make changes to the calling convention lowering code and possibly use addRegisterClass for f16.

Reply:
I noticed that RISC-V introduced fp16 register class specifically for the Zfh extension. I’m wondering if it would be better to wait until the formal fp16 extension is supported in the future before doing something similar. For now, software fp16 is being handled using the same approach as RISC-V to ensure compatibility with future hardware fp16. Thanks.

Reply:
@asb What do you think? Thanks.

Reply:
You don't need a new register class to pass in an existing register class.

You can add f16 to the GPR register class and then make all operations expand, or adjust the calling convention code to bitcast to i16, or whatever will get promoted into gps

Reply:
I don't understand why are GPR and `i16`. I think this custom lowering ensures that the argument type of makeLibCall is `f32`, not an integer. If I do this, how can I move the argument of `__gnu_f2h_ieee` to the FPR? Thanks.

Reply:
Whatever the classes and types are, I know nothing about this architecture. You need to change the code that determines what register class ends up used in ABI contexts. 

Look at computeRegisterProperties and getRegisterTypeForCallingConv. You need to adjust some combination of the registered legal types, and the calling convention code in the *CallingConv.td files, and possibly any custom code in LowerCall, LowerReturn LowerFormalArguments, and getRegisterTypeForCallingConv

Reply:
For this case:

```llvm
define float @test_fpextend_float(half %a) nounwind {
  %r = fpext half %a to float
  ret float %r
}
```
Debug log (without FP16_TO_FP custom lowering):

```
Optimized lowered selection DAG: %bb.0 'test_fpextend_float:'
SelectionDAG has 10 nodes:
  t0: ch,glue = EntryToken
            t2: f32,ch = CopyFromReg t0, Register:f32 %0
          t3: i32 = bitcast t2
        t4: i16 = truncate t3
      t5: f16 = bitcast t4
    t6: f32 = fp_extend t5
  t8: ch,glue = CopyToReg t0, Register:f32 $f0, t6
  t9: ch = LoongArchISD::RET t8, Register:f32 $f0, t8:1

Type-legalized selection DAG: %bb.0 'test_fpextend_float:'
SelectionDAG has 10 nodes:
  t0: ch,glue = EntryToken
          t2: f32,ch = CopyFromReg t0, Register:f32 %0
        t10: i64 = LoongArchISD::MOVFR2GR_S_LA64 t2
      t14: i64 = and t10, Constant:i64<65535>
    t12: f32 = fp16_to_fp t14
  t8: ch,glue = CopyToReg t0, Register:f32 $f0, t12
  t9: ch = LoongArchISD::RET t8, Register:f32 $f0, t8:1

Optimized type-legalized selection DAG: %bb.0 'test_fpextend_float:'
SelectionDAG has 8 nodes:
  t0: ch,glue = EntryToken
        t2: f32,ch = CopyFromReg t0, Register:f32 %0
      t10: i64 = LoongArchISD::MOVFR2GR_S_LA64 t2
    t15: f32 = fp16_to_fp t10
  t8: ch,glue = CopyToReg t0, Register:f32 $f0, t15
  t9: ch = LoongArchISD::RET t8, Register:f32 $f0, t8:1

Legalized selection DAG: %bb.0 'test_fpextend_float:'
SelectionDAG has 8 nodes:
  t0: ch,glue = EntryToken
      t2: f32,ch = CopyFromReg t0, Register:f32 %0
    t10: i64 = LoongArchISD::MOVFR2GR_S_LA64 t2
  t18: ch,glue = CopyToReg t0, Register:i64 $r4, t10
  t20: ch,glue = LoongArchISD::TAIL t18, TargetExternalSymbol:i64'__gnu_h2f_ieee' [TF=2], Register:i64 $r4, t18:1
```

In joinRegisterPartsIntoValue, the fp16 (wrapped as f32) in the parameter register (FPR) is first bitcast to i32, truncated to i16, and then bitcast to f16. Following this, the software promotion of FP_EXTEND generates FP16_TO_FP. At this point, the source operand of FP16_TO_FP is already i64, losing its association with the fp16 type. It seems that the target calling convention cannot handle passing through the FPR? :sob:

Reply:
By default the half is going to be be pre-promoted to a legal f32 type and you need to intervene before that. I would start by overriding getRegisterTypeForCallingConv, and then see how that goes. You may need to just custom hack on the argument lists in each of the Lower* functions. 

Yes SelectionDAG makes this more difficult than it should be. It would be much easier if calling convention code operated on the raw IR signature instead of going through type legalization first

Reply:
Thanks for your guidance. I haven't quite managed to get it done yet. To pass fp16 in FPR (excluding libcall), `splitValueIntoRegisterParts` and `joinRegisterPartsIntoValue` insert all ones in the upper bits and extract the lower bits by casting to an integer type. I guess this is the key point where the fp16 value is promoted to an integer when it reaches the `FP16_TO_FP` operation, but I'm not sure how to bypass the integer type to achieve this.

Additionally, it seems that custom-lowering `FP16_TO_FP` and `FP_TO_FP16` to generate a libcall while keeping it passed in FPR works quite well and is fairly easy to implement, RISC-V is already using this approach. Can we go ahead with this? Thanks.

Reply:
It's not really correct. It only happens to work out, and MOVFR2GR_S_LA64 is a hack

Reply:
> In Rust's implementation, the argument of `__gnu_h2f_ieee` is `f16`, not `i16`.
> 
> https://github.com/rust-lang/compiler-builtins/blob/compiler_builtins-v0.1.126/src/float/extend.rs#L95-L97
> 
> ```rust
> pub extern "C" fn __gnu_h2f_ieee(a: f16) -> f32 {
>     extend(a)
> }
> ```

Just to address this (quite late) - aiui `__gnu_h2f_ieee` is only called on platforms where `f16` is passed as an integer so that doesn't matter here. If there is a float ABI then `__extendhfsf2`/`__truncsfhf2` is used instead.

There is something weird with the return value though, Rust's compiler-builtins and LLVM's compiler-rt both use `f32` but GCC uses `int`. (I think maybe GCC just never emits that libcall except on ARM, LLVM seems to use it a lot more).

Reply:
In my view, `__gnu_h2f_ieee` is primarily designed for converting `f16` values to `f32`, particularly in scenarios where hardware lacks native `f16` support, requiring software emulation instead. This function does not inherently define any specific argument-passing conventions; rather, those are determined by the architecture's ABI. For instance, on RISC-V, the lower 16 bits of a floating point argument register are used (given that hardware support for `f16` is not enabled), while some other architectures use integer registers. In terms of implementation, `__extendhfsf2` serves as an alias for `__gnu_h2f_ieee` and are used by arm.

https://github.com/rust-lang/compiler-builtins/blob/compiler_builtins-v0.1.126/src/float/extend.rs#L87-L89

```rust
pub extern "C" fn __extendhfsf2(a: f16) -> f32 {
    extend(a)
}
```

https://github.com/rust-lang/compiler-builtins/blob/compiler_builtins-v0.1.126/src/float/extend.rs#L95-L97

```rust
pub extern "C" fn __gnu_h2f_ieee(a: f16) -> f32 {
    extend(a)
}
```